feat(repository): add helpers to create relations

jimmy1wu opened the pull request:
* create helpers for belongsTo, hasOne, and hasMany relations
* infer the keyFrom value for belongsTo relation

Fixes https://github.com/strongloop/loopback-next/issues/5384

<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

-->

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

agnes512 commented on packages/repository/src/relations/belongs-to/belongs-to.helpers.ts:
```diff
@@ -49,12 +48,24 @@ export function resolveBelongsToMetadata(relationMeta: BelongsToDefinition) {
   const targetName = targetModel.modelName;
   debug('Resolved model %s from given metadata: %o', targetName, targetModel);
 
+  let keyFrom;
+  if (
+    relationMeta.keyFrom &&
+    relationMeta.source.definition.properties[relationMeta.keyFrom]
+  ) {
+    keyFrom = relationMeta.keyFrom;
+  } else {
+    keyFrom = camelCase(targetName + '_id');
+  }
+
```
I think the current design allows users to customize the `keyFrom` and `keyTo` value. And `keyFrom` is default to the decorated key (e.g `customerId`), see [code](https://github.com/strongloop/loopback-next/blob/master/packages/repository/src/relations/belongs-to/belongs-to.decorator.ts#L50). Even it throws when `keyFrom` is undefined, I don't think this situation would happen as it would default the decorated key anyway. So I don't think the change is necessary.

bajtos replied:
I think the intention here is to allow users to omit `keyFrom` when calling `modelDefinition.belongsTo()` - that's a different code path from the one used by `@belongsTo` decorator. When calling `belongsTo`, I think it's useful to be able to leave out `keyFrom` and let the framework to compute it.

agnes512 commented on packages/repository/src/model.ts:
```diff
@@ -145,6 +149,63 @@ export class ModelDefinition {
     return this;
   }
 
+  /**
```
What't the intention of adding these helpers? And I think relations should not be the part of the model definition.

bajtos commented on packages/repository/src/model.ts:
```diff
@@ -145,6 +149,63 @@ export class ModelDefinition {
+  belongsTo(
+    name: string,
+    definition: Partial<BelongsToDefinition> &
+      Pick<BelongsToDefinition, 'source' | 'target'>,
```
I find it a bit difficult to understand which properties are required and which are optional. Relation definitions have most properties optional, do we need to apply `Partial` here?

It makes me wonder if something like the following would work too:

```ts
definition: Omit<BelongsToDefinition, 'name' | 'type' | 'targetsMany'>,
```

Feel free to ignore this comment if my suggestion requires too many changes to make it work.

jimmy1wu replied:
I agree this is a lot more clear. I will update the PR soon.

bajtos commented on packages/repository/src/model.ts:
```diff
@@ -145,6 +149,63 @@ export class ModelDefinition {
+      ...definition,
+      name,
+      type: RelationType.belongsTo,
+      targetsMany: false,
```
Is it possible to infer `keyFrom` value from the target model name? Let's say `target` is `() => Customer` and we are defining the relation "order belongs to a customer", would something like this work?

```ts
if (! belongsToDefinition.keyFrom) {
  const targetName = definition.target().modelName;
  // ^^ should be "Customer"
  belongsToDefinition.keyFrom = camelCase(`${targetName}_Id`);
  // ^^ should produce "customerId"
}
```

A similar comment applies to other relation types too - `hasOne` and `hasMany` can use `definition.source.modelName` to build the foreign key name.

Since this is an incremental addition, I am ok to leave it out of this initial version if you prefer.

bajtos replied:
Oh, I see that you are computing `keyFrom` in `resolveBelongsToMetadata`, that's a good solution too 👍

jimmy1wu replied:
I tried your suggestion and adding the logic inside the belongsTo function may lead to reference errors if they define the target class after the class that is using the belongsTo method, so adding the logic in `resolveBelongsToMetadata` is more flexible.

e.g.,
https://github.com/strongloop/loopback-next/blob/master/packages/repository/src/__tests__/integration/repositories/relation.factory.integration.ts#L464

The `Customer` class variable is available in the scope, but has not been initialized yet so it will throw a reference error if we try to access it.

bajtos commented on packages/repository/src/relations/relation.types.ts:
```diff
@@ -117,7 +117,7 @@ export interface BelongsToDefinition extends RelationDefinitionBase {
    * The foreign key in the source model, e.g. Order#customerId.
    */
-  keyFrom: string;
+  keyFrom?: string;
```
I have mixed feelings about this change. On one hand, I think it allows us to simplify `belongsTo` method to accept `definition: Omit<BelongsToDefinition, 'name' | 'type' | 'targetsMany'>` and is kind of consistent with other relation types that have `keyFrom` optional too.

On the other hand, I am not able to fully asses ramifications of this change, I have no idea what it could break.

On the third hand :) Since this is backwards compatible and you have code to handle the case where `keyFrom` is not provided, I think it should be fine.

bajtos commented on packages/repository/src/model.ts:
```diff
@@ -145,6 +149,63 @@ export class ModelDefinition {
+   */
+  hasOne(
+    name: string,
+    definition: Partial<HasOneDefinition> &
+      Pick<HasOneDefinition, 'source' | 'target'>,
```
Similarly here, would the following work?

```suggestion
    definition: Omit<HasOneDefinition, 'name' | 'type' | 'targetsMany'>,
```

bajtos commented on packages/repository/src/model.ts:
```diff
@@ -145,6 +149,63 @@ export class ModelDefinition {
+   */
+  hasMany(
+    name: string,
+    definition: Partial<HasManyDefinition> &
+      Pick<HasManyDefinition, 'source' | 'target'>,
```
ditto

```suggestion
    definition: Omit< HasManyDefinition, 'name' | 'type' | 'targetsMany'>,
```

jimmy1wu commented:
@agnes512 @bajtos 
Thank you for the early review and suggestions :)
I will update the PR shortly.

Looking into tests today / this weekend

bajtos commented on packages/repository/src/relations/belongs-to/belongs-to.helpers.ts:
```diff
@@ -49,12 +48,24 @@ export function resolveBelongsToMetadata(relationMeta: BelongsToDefinition) {
+  ) {
+    keyFrom = relationMeta.keyFrom;
+  } else {
+    keyFrom = camelCase(targetName + '_id');
```
Please add a test case to execute this line.

bajtos commented on packages/repository/src/model.ts:
```diff
@@ -145,6 +149,60 @@ export class ModelDefinition {
+   * @param name - The name of the "belongs to" relation.
+   * @param definition - The definition for the "belongs to" relation.
+   */
+  belongsTo(
```
Please add a test to verify this new API.

bajtos commented on packages/repository/src/model.ts:
```diff
@@ -145,6 +149,60 @@ export class ModelDefinition {
+   * @param name - The name of the "has one" relation.
+   * @param definition - The definition of the "has one" relation.
+   */
+  hasOne(
```
Please add a test to verify this new API.

bajtos commented on packages/repository/src/model.ts:
```diff
@@ -145,6 +149,60 @@ export class ModelDefinition {
+   * @param name - The name of the "has many" relation.
+   * @param definition - The definition of the "has many" relation.
+   */
+  hasMany(
```
Please add a test to verify this new API.

jimmy1wu commented:
@agnes512 ready for review 😁

agnes512 commented on packages/repository/src/__tests__/unit/model/model.unit.ts:
```diff
@@ -313,11 +307,7 @@ describe('model', () => {
     Category.definition = new ModelDefinition('Category')
       .addSetting('strict', false)
       .addProperty('id', {type: 'number', id: true, required: true})
-      .addRelation(<HasManyDefinition>{
```
Let's revert this part and the test at line 479, or there is no tests cover `addRelation`.

agnes512 commented on packages/repository/src/__tests__/unit/model/model.unit.ts:
```diff
@@ -508,4 +495,66 @@ describe('model', () => {
     });
   });
+
+  describe('test adding relations through helpers', () => {
```
Nitpick: `describe('relation helpers', ..`

jimmy1wu commented:
@agnes512 
while addressing your feedback, i noticed that i wasn't always importing from the common index file and added a commit to fix it. sorry about that; trusted my IDE too much! 

the rest of your comments are addressed.

agnes512 commented on packages/repository/src/__tests__/unit/model/model.unit.ts:
```diff
@@ -263,11 +263,7 @@ describe('model', () => {
     Category.definition = new ModelDefinition('Category')
       .addSetting('strict', true)
       .addProperty('id', {type: 'number', id: true, required: true})
-      .addRelation(<HasManyDefinition>{
```
Please revert this part as well.
